src/librustc_typeck/coherence/overlap.rs

```diff
@@ -48,28 +48,28 @@ impl<'cx, 'tcx> OverlapChecker<'cx, 'tcx> {
             Value,
         }
 
-        let name_and_namespace = |def_id| {
-            let item = self.tcx.associated_item(def_id);
+        let name_and_namespace = |r: &ty::AssociatedItemRef| {
+            let item = self.tcx.associated_item(r.def_id);
```

Comment:
I think I agree. I'd almost go as far as having `associated_item` use the `DefId` of its container for the dep node, and then just use `associated_items` everywhere.